File: backend/api/src/main/kotlin/me/ricoschulz/crawler/Asmodee.kt
package me.ricoschulz.crawler

import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import me.ricoschulz.helper.fetch
import me.ricoschulz.helper.hasKeyword
import me.ricoschulz.helper.json
import me.ricoschulz.helper.selectTags
import me.ricoschulz.valueObject.News
import me.ricoschulz.valueObject.Product
import me.ricoschulz.valueObject.Provider
import me.ricoschulz.valueObject.allNames
import org.jsoup.Jsoup

@Serializable internal data class SmallDto(val url: String)

@Serializable internal data class FormatsDto(val small: SmallDto)

@Serializable
internal data class TileImageDto(val formats: FormatsDto? = null, val url: String? = null)

@Serializable internal data class FacetsDto(val image: String)

@Serializable
internal data class ProductDto(val name: String, val facets: FacetsDto, val slug: String) {
    fun toProductDto() =
        Product(name = name, img = facets.image, link = "https://www.asmodee.de/produkte/${slug}")
}

@Serializable
internal data class ContentDto(val text: String? = null, val product: ProductDto? = null)

@Serializable
internal data class ArticleDto(
    val headline: String,
    val subHeadline: String,
    val creationDate: String,
    val tileImage: TileImageDto,
    val slug: String,
    val content: List<ContentDto> = listOf()
) {
    fun product(contentFilter: (ContentDto) -> Boolean) =
        this.content.filter(contentFilter).map { it.product!!.toProductDto() }

    fun image() = tileImage.formats?.small?.url ?: tileImage.url ?: throw Throwable("missing image")
}

@Serializable internal data class PagePropsDto(val articles: List<ArticleDto>)

@Serializable
internal data class AsmodeeResponseDto(val pageProps: PagePropsDto) {
    fun articleForEach(articleFilter: (ArticleDto) -> Boolean, callback: (ArticleDto) -> Unit) {
        pageProps.articles.filter(articleFilter).forEach(callback)
    }
}

@DelicateCoroutinesApi
suspend fun asmodee(callback: (News) -> Unit) =
    GlobalScope.async {
        fun contentFilter(content: ContentDto) =
            content.product != null && content.product.name.hasKeyword()

        fun articleFilter(article: ArticleDto) =
            article.headline.plus(article.subHeadline).hasKeyword() ||
                article.content.any(::contentFilter)

        suspend fun loadHash(): String {
            val (hash) =
                "/_next/static/(.*)/_buildManifest.js"
                        .toRegex()
                        .find(
                            Jsoup.parse(fetch("https://www.asmodee.de"))
                                .select("script[src$=_buildManifest.js]")
                                .attr("src")
                        )!!
                    .destructured
            return hash
        }

        val response: String = fetch("https://www.asmodee.de/_next/data/${loadHash()}/news.json")
        json.decodeFromString<AsmodeeResponseDto>(response).articleForEach(::articleFilter) {
            val products = it.product(::contentFilter)

            callback(
                News(
                    title = it.headline,
                    newsType = Provider.ASMODEE,
                    subTitle = it.subHeadline,
                    link = "https://www.asmodee.de/news/${it.slug}",
                    img = it.image(),
                    date = it.creationDate,
                    products = products,
                    tags = selectTags(products.allNames(), "Asmodee")
                )
            )
        }
    }

File: server/src/Domains/Article/Services/Provider/Asmodee.php
<?php

declare(strict_types=1);

namespace Domains\Article\Services\Provider;

use Carbon\CarbonImmutable;
use Domains\Article\Aggregates\Article;
use Domains\Article\Entities\Product;
use Domains\Article\Exceptions\ProviderException;
use Domains\Article\Factories\ArticleFactory;
use Domains\Article\Factories\ProductFactory;
use Domains\Article\Repositories\Articles;
use Domains\Article\Services\Crawler;
use Domains\Article\Services\Crawler\Filter\KeywordFilter;
use Domains\Article\Services\Crawler\Filter\ProductExistFilter;
use Domains\Article\Services\HttpClient;
use Domains\Article\ValueObjects\Provider;

class Asmodee implements Crawler
{
    private const ASMODEE_HOME_URL = 'https://www.asmodee.de';
    private const ASMODEE_NEWS_URL = 'https://www.asmodee.de/_next/data/%s/news.json';

    public function __construct(
        private readonly KeywordFilter $keywordFilter,
        private readonly ProductExistFilter $productExistFilter,
        private readonly Articles $articles,
        private readonly HttpClient $httpClient,
        private readonly ArticleFactory $articleFactory,
        private readonly ProductFactory $productFactory,
    ) {}

    public function crawl(): void
    {
        $newUrl = $this->loadNewsUrl();
        $response = $this->httpClient->loadContentFromWebsite($newUrl);

        $allArticles = $this->mapArrayToArticle($response['pageProps']['articles']);
        $filterArticles = $this->keywordFilter->removeArticlesWhoNotMatchAnyKeyword($allArticles);

        $this->articles->addAll(...$filterArticles);
    }

    private function loadNewsUrl(): string
    {
        $dom = $this->httpClient->loadContentFromWebsite(self::ASMODEE_HOME_URL);
        $scripts = $dom->getElementsByTagName('script');

        foreach ($scripts as $script) {
            if (1 === preg_match('/\/_next\/static\/(.*?)\/_buildManifest.js/', $script->getAttribute('src'), $match)) {
                return sprintf(self::ASMODEE_NEWS_URL, $match[1]);
            }
        }

        throw new ProviderException('No hash found for Asmodee news API call!');
    }

    private function mapArrayToArticle(array $allArticleData): array
    {
        $emptyProductContentFilter = static fn (array $productContent): bool => false === empty($productContent['product']);

        $mapping = function (array $articleData) use ($emptyProductContentFilter): Article {
            $contents = $articleData['content'] ?? [];
            $contentWithProduct = array_filter($contents, $emptyProductContentFilter);
            $allProducts = $this->createProductFromArray($contentWithProduct);
            $filteredProducts = $this->productExistFilter->removeKnownProducts($allProducts);

            return $this->createArticleFromArray($articleData, $filteredProducts);
        };

        return array_map($mapping, $allArticleData);
    }

    private function createArticleFromArray(array $articleData, array $products): Article
    {
        $linkCategory = empty($products) ? 'product' : 'news';

        return $this->articleFactory->setArticleData(
            provider: Provider::ASMODEE,
            headline: $articleData['headline'],
            publishDate: CarbonImmutable::parse($articleData['creationDate']),
            image: $articleData['tileImage']['formats']['small']['url'] ?? $articleData['tileImage']['url'],
            link: "https://www.asmodee.de/{$linkCategory}/{$articleData['slug']}",
            subHeadline: $articleData['subHeadline'],
        )->setProducts($products)
            ->build()
        ;
    }

    private function createProductFromArray(array $contentDatas): array
    {
        $mapping = fn (array $contentData): Product => $this->productFactory->setProductData(
            name: $contentData['product']['name'],
            link: 'https://www.asmodee.de/produkte/'.$contentData['product']['slug'],
            image: $contentData['product']['facets']['image'],
        )->build();

        return array_map($mapping, $contentDatas);
    }
}
